### PR TITLE
RORD eraser: scaffold

### PR DESCRIPTION
**Pull request by piercus**
First version including : 
* Neptune logger
* Basic README.md
* SD 1.5 related changes

Not including : 
* Full reference metrics (PIQ-like)
* Aspect ratio bucketing/pixel budget
* Eraser-tailored masking strategy

To see the diff with lbm_surface, compare to d3a9a8b57cb62dd319d20110097903e5f34e4a62

### CODE REVIEW

File: contrib/eraser/README.md [hunk 1]
```diff
@@ -0,0 +1,44 @@
+## LBM Eraser
+
+Reproduce Eraser result from LBM paper on SD1.5.
```
Comment (deltheil) [inline contrib/eraser/README.md]:
```suggestion
**Goal**: reproduce the object-removal result from the LBM paper (Section 4.1) using SD1.5.
```

File: contrib/eraser/README.md [hunk 2]
```diff
@@ -0,0 +1,44 @@
...
+Non-official implementation.
+
+### Dataset
+
+Prepare a group [webdataset](https://github.com/webdataset/webdataset) shards organized like
```
Comment (deltheil) [inline contrib/eraser/README.md]:
```suggestion
Prepare a group of [webdataset](https://github.com/webdataset/webdataset) shards, organized like:
```

File: contrib/eraser/README.md [hunk 3]
```diff
@@ -0,0 +1,44 @@
...
+
+```
+<id0>.after.jpg
+<id0>.before.jpg
+<id0>.mask.png
+<id1>.after.jpg
+<id1>.before.jpg
+<id1>.mask.png
+...
+```
```
Comment (deltheil) [inline contrib/eraser/README.md:18]:
```suggestion
```

Where "before" is the image containing the object to remove, "mask" is the (fine) segmentation mask of the object in the before image, and "after" is the image without that object (ground truth).

```

File: contrib/eraser/README.md [hunk 4]
```diff
@@ -0,0 +1,44 @@
...
+
+### Train
+
+To train the model on a single GPU machine, you can use the following command.
+Works on a single RTX 3090
```
Comment (deltheil) [inline contrib/eraser/README.md]:
```suggestion
Works on a single RTX 3090:
```

File: contrib/eraser/README.md [hunk 5]
```diff
@@ -0,0 +1,44 @@
...
+export SLURM_JOB_ID=0000 
+export SLURM_PROCID=0 
+export SLURM_ARRAY_TASK_ID=0
+export SLURM_NPROCS=1
+export SLURM_NODEID=0
+export SLURM_NNODES=1
+export SLURM_LOCALID=0 
+
+# neptune mode, 
+export NEPTUNE_MODE=async # use 'offline' to disable neptune logging
```
Comment (deltheil) [inline contrib/eraser/README.md]:
```suggestion
# Neptune (https://neptune.ai) settings. Use 'offline' mode to disable logging.
export NEPTUNE_MODE=async
```

File: examples/training/config/eraser.yaml [hunk 6]
```diff
@@ -0,0 +1,31 @@
...
+
+# SHARDS_PATH_OR_URLS
+train_shards:
+  - pipe:cat $PATH_TO_TRAIN_TARS
+
+validation_shards:
+  - pipe:cat $PATH_TO_VAL_TARS
+
+batch_size: 4
+learning_rate: 4e-5
```
Comment (deltheil) [inline examples/training/config/eraser.yaml]: Any reason not to pick 3e-5 (to start with) like in appendix A.1?

File: examples/training/config/eraser.yaml [hunk 7]
```diff
@@ -0,0 +1,31 @@
...
+timestep_sampling: custom_timesteps
+unet_input_channels: 4
+vae_num_channels: 4
+selected_timesteps: [250, 500, 750, 1000]
+prob: [0.25, 0.25, 0.25, 0.25]
+pixel_loss_type: lpips # l1 l2
+pixel_loss_weight: 10.0
+latent_loss_type: l2 # l1 l2
+latent_loss_weight: 1.0
+bridge_noise_sigma: 0.005
```
Comment (deltheil) [inline examples/training/config/eraser.yaml]: Why not 0.05 like in appendix A.1?

File: contrib/eraser/training/config/eraser.yaml [hunk 8]
```diff
@@ -0,0 +1,31 @@
...
+conditioning_masks_keys: []
+
+# SHARDS_PATH_OR_URLS
+train_shards:
+  - pipe:cat $PATH_TO_TRAIN_TARS
+
+validation_shards:
+  - pipe:cat $PATH_TO_VAL_TARS
+
+batch_size: 4
```
Comment (deltheil) [inline contrib/eraser/training/config/eraser.yaml:24]: I guess you don't support budgets yet?

Image

File: contrib/eraser/training/train_eraser.py [hunk 9]
```diff
@@ -0,0 +1,550 @@
...
+
+    # Load pretrained model as base
+    pipe = StableDiffusionPipeline.from_pretrained(
+        backbone_signature,
+        torch_dtype=torch.bfloat16,
+    )
+
+    ### MMMDiT ###
+    # Get Architecture
+    denoiser = DiffusersUNet2DCondWrapper(
```
Comment (deltheil) [inline contrib/eraser/training/train_eraser.py:78]: Just to confirm you did a side-by-side comparison with https://huggingface.co/stable-diffusion-v1-5/stable-diffusion-v1-5/blob/main/unet/config.json

File: contrib/eraser/training/train_eraser.py [hunk 10]
```diff
@@ -0,0 +1,550 @@
...
+    state_dict_adapter = StateDictAdapter()
+    state_dict = state_dict_adapter(
+        model_state_dict=denoise_state_dict,
+        checkpoint_state_dict=state_dict,
+        regex_keys=[
+            r"class_embedding.linear_\d+.(weight|bias)",
+            r"conv_in.weight",
+            r"(down_blocks|up_blocks)\.\d+\.attentions\.\d+\.transformer_blocks\.\d+\.attn\d+\.(to_k|to_v)\.weight",
+            r"mid_block\.attentions\.\d+\.transformer_blocks\.\d+\.attn\d+\.(to_k|to_v)\.weight",
+        ],
```
Comment (deltheil) [inline contrib/eraser/training/train_eraser.py:154]: Just to confirm: these (made for SDXL UNet) work verbatim with SD1.5, right?

File: examples/training/train_eraser.py [hunk 11]
```diff
@@ -0,0 +1,550 @@
...
+    conditioner = ConditionerWrapper(
+        conditioners=conditioners,
+    )
+
+    ## VAE ##
+    # Get VAE model
+    vae_config = AutoencoderKLDiffusersConfig(
+        version=backbone_signature,
+        subfolder="vae",
+        tiling_size=(128, 128),
```
Comment (deltheil) [inline examples/training/train_eraser.py]: Maybe this can be reduced (i.e. just keep the default which is 64)? Anyway, we probably won't use tiling for now

File: examples/training/train_eraser.py [hunk 12]
```diff
@@ -0,0 +1,550 @@
...
+                    )
+                ),
+                TorchvisionMapper(
+                    TorchvisionMapperConfig(
+                        key="before",
+                        transforms=["ToTensor", "Resize"],
+                        transforms_kwargs=[
+                            {},
+                            {
+                                "size": (480, 640),
```
Comment (deltheil) [inline examples/training/train_eraser.py]: Not set in stone, right?

File: examples/training/train_eraser.py [hunk 13]
```diff
@@ -0,0 +1,550 @@
...
+def main(
+    train_shards: List[str] = ["pipe:cat path/to/train/shards"],
+    validation_shards: List[str] = ["pipe:cat path/to/validation/shards"],
+    backbone_signature: str = "stable-diffusion-v1-5/stable-diffusion-v1-5",
+    vae_num_channels: int = 4,
+    unet_input_channels: int = 4,
+    source_key: str = "before",
+    target_key: str = "after",
+    mask_key: str = "mask",
+    neptune_project: str = "LBM Eraser",
```
Comment (deltheil) [inline examples/training/train_eraser.py]: nit: you've used "LBM-Eraser" in `eraser.yaml` (maybe keep the same default here)

File: examples/training/train_eraser.py [hunk 14]
```diff
@@ -0,0 +1,550 @@
...
+            "config_yaml": config_yaml,
+            "training": training_config.to_dict(),
+            "training_noise_scheduler": model.training_noise_scheduler.config,
+            "sampling_noise_scheduler": model.sampling_noise_scheduler.config,
+        }
+    )
+
+    training_signature = (
+        datetime.datetime.now().strftime("%Y-%m-%d_%H-%M-%S")
+        + "-LBM-Surface"
```
Comment (deltheil) [inline examples/training/train_eraser.py]:
```suggestion
        + "-LBM-Eraser"
```

File: src/lbm/trainer/loggers.py [hunk 15]
```diff
@@ -322,3 +322,118 @@ def _process_logs(
                 )
 
         return logs
+
+
+class NeptuneLogger(Callback):
```
Comment (deltheil) [inline src/lbm/trainer/loggers.py:327]: What is specific (i.e. differs) here compared to `WandbSampleLogger` (didn't check)? So the `trainer.logger` is different (ofc).

**Review by deltheil — COMMENTED** [review]
Some nit-s and questions. PTAL

File: src/lbm/trainer/loggers.py (hunk 15 above)
Comment (deltheil) [reply src/lbm/trainer/loggers.py:327]:
```diff
--- a.py	2025-10-08 18:38:36
+++ b.py	2025-10-08 18:39:12
@@ -1,9 +1,9 @@
-class WandbSampleLogger(Callback):
+class NeptuneLogger(Callback):
     """
-    Logger for logging samples to wandb. This logger is used to log images, text, and metrics to wandb.
+    Logger for logging samples to Neptune. This logger is used to log images, text, and metrics to Neptune.
 
     Args:
-        log_batch_freq (int): The frequency of logging samples to wandb. Default is 100.
+        log_batch_freq (int): The frequency of logging samples to Neptune. Default is 100.
     """
 
     def __init__(self, log_batch_freq: int = 100):
@@ -56,7 +56,7 @@
                     pl_module.train()
         else:
             logging.warning(
-                "log_img method not found in LightningModule. Skipping image logging."
+                "log_samples method not found in LightningModule. Skipping image logging."
             )
 
     @rank_zero_only
@@ -74,26 +74,25 @@
                     grid = grid.permute(1, 2, 0)
                     grid = grid.mul(255).clamp(0, 255).to(torch.uint8)
                     logs[key] = grid.numpy()
-                    trainer.logger.experiment.log(
-                        {f"{key}/{split}": [wandb.Image(Image.fromarray(logs[key]))]},
+                    trainer.logger.experiment[f"{key}/{split}"].append(
+                        NeptuneFile.as_image(Image.fromarray(logs[key])),
                         step=trainer.global_step,
                     )
 
                 # Scalar tensor
                 if value.dim() == 1 or value.dim() == 0:
                     value = value.float().numpy()
-                    trainer.logger.experiment.log(
-                        {f"{key}/{split}": value}, step=trainer.global_step
+                    trainer.logger.experiment[f"{key}/{split}"].append(
+                        value, step=trainer.global_step
                     )
 
             # list of string (e.g. text)
             if isinstance(value, list):
                 if isinstance(value[0], str):
                     pil_image_texts = create_grid_texts(value)
-                    wandb_image = wandb.Image(pil_image_texts)
-                    trainer.logger.experiment.log(
-                        {f"{key}/{split}": [wandb_image]},
-                        step=trainer.global_step,
+                    neptune_image = NeptuneFile.as_image(Image.fromarray(pil_image_texts))
+                    trainer.logger.experiment[f"{key}/{split}"].append(
+                        neptune_image, step=trainer.global_step
                     )
 
             # dict of tensors (e.g. metrics)
@@ -101,13 +100,14 @@
                 for k, v in value.items():
                     if isinstance(v, torch.Tensor):
                         value[k] = v.detach().cpu().numpy()
-                trainer.logger.experiment.log(
-                    {f"{key}/{split}": value}, step=trainer.global_step
+
+                trainer.logger.experiment[f"{key}/{split}"].append(
+                    value, step=trainer.global_step
                 )
 
             if isinstance(value, int) or isinstance(value, float):
-                trainer.logger.experiment.log(
-                    {f"{key}/{split}": value}, step=trainer.global_step
+                trainer.logger.experiment[f"{key}/{split}"].append(
+                    value, step=trainer.global_step
                 )
 
-        return logs
\ No newline at end of file
+        return logs
```

File: contrib/eraser/training/train_eraser.py (hunk 10 above)
Comment (piercus) [reply contrib/eraser/training/train_eraser.py:154]: Yes the main difference is that we do not need to delete `add_embbeding.*` (see [here](https://github.com/finegrain-ai/LBM/blob/cafebc46a9ac16dcc61691d289cc4676b5c75380/examples/training/train_lbm_surface.py#L134-L137))

File: contrib/eraser/training/train_eraser.py (hunk 9 above)
Comment (piercus) [reply contrib/eraser/training/train_eraser.py:78]: Yes, `cross_attention_dim` is the main difference, because the cross-attentions are repurposed and zeroed as self-attentions.

File: contrib/eraser/training/config/eraser.yaml (hunk 8 above)
Comment (piercus) [reply contrib/eraser/training/config/eraser.yaml:24]: Yes the aspect ratio is not included in this PR

File: examples/training/train_eraser.py (hunk 11 above)
Comment (piercus) [reply examples/training/train_eraser.py]: tiling is currently not activated, yes i will remove this for clarity

Comment (deltheil) [reply examples/training/train_eraser.py]: OK. That said, it will depend on the input image size (IIUC).

File: examples/training/train_eraser.py (hunk 12 above)
Comment (piercus) [reply examples/training/train_eraser.py]: added image_size in `eraser.yaml`

**Review by deltheil — APPROVED** [review]
LGTM for this initial scaffold
